### PR TITLE
CPLAT-7608 [3.1.0] Improve handling of “repeat” errors thrown from components wrapped by an ErrorBoundary

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
+ 3.1.0-wip port of #350

> See #350 for full description

@greglittlefield-wf @kealjones-wk @joebingham-wk @sydneyjodon-wk

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
The items listed below may not capture all security relevant changes. Before providing a security review, be sure to review the entire PR for security impact.
<details>
<summary>(3) Security relevant changes were detected</summary>
<li>Watched keyword <code>innerHtml</code> in <code>lib/src/component/error_boundary_mixins.dart</code> line(s) ['258'] added</li>
<li>Watched keyword <code>InnerHtml</code> in <code>lib/src/component/error_boundary_mixins.dart</code> line(s) ['282'] added</li>
<li>Watched keyword <code>dangerouslySetInnerHTML</code> in <code>lib/src/component/error_boundary_mixins.dart</code> line(s) ['283'] added</li>
</details>

## Action Items
* Obtain a [security review](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); reviewer should pay special attention to insights listed above
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/component/error_boundary_mixins.dart [hunk 1]
```diff
@@ -78,30 +141,174 @@ abstract class _$ErrorBoundaryStateMixin implements UiState {
...
-    if (state.hasError && props.fallbackUIRenderer != null) {
-      return props.fallbackUIRenderer();
+    if (state.hasError && state.showFallbackUIOnError) {
+      return (props.fallbackUIRenderer ?? _renderStringDomAfterUnrecoverableErrors)(_lastError, _lastErrorInfo);
     }
 
     return props.children;
   }
+
+  // TODO: Add PropTypes
```
Comment (aaronlademann-wf) [inline lib/src/component/error_boundary_mixins.dart]: Not in this PR... I just noticed that `validateProps` had been removed without having an analogous propTypes replacement.

**Comment by maxwellpeterson-wf** [comment]
+1 security
- Our CSP protects us from any potential XSS made possible by this change.

File: test/over_react/component/fixtures/flawed_component.dart [hunk 2]
```diff
@@ -32,28 +35,56 @@ class FlawedState extends _$FlawedState with _$FlawedStateAccessorsMixin {
 @Component()
 class FlawedComponent extends UiStatefulComponent<FlawedProps, FlawedState> {
```
Comment (joebingham-wk) [inline test/over_react/component/fixtures/flawed_component.dart]: I know this matches the `3.1.0` branch implementation and was `Component` to begin with, but was there a reason it didn't get updated to `Component2` here?

File: test/over_react/component/shared_error_boundary_tests.dart [hunk 3]
```diff
@@ -102,26 +114,605 @@ void sharedErrorBoundaryTests(BuilderOnlyUiFactory builder) {
...
+
+            expect(errorSentToComponentDidCatchCallback, isA<FlawedComponentException2>());
+            expect(errorInfoSentToComponentDidCatchCallback, isA<ReactErrorInfo>());
+          });
+        });
+      });
     });
   });
+
+//  group('throws a PropError when', () {
```
Comment (joebingham-wk) [inline test/over_react/component/shared_error_boundary_tests.dart]: Should we add a TODO here to mention that the validate props needs to be added back?

**Comment by kealjones-wk** [comment]
+10

**Comment by kealjones-wk** [comment]
@Workiva/release-management-p

**Comment by kealjones-wk** [comment]
@Workiva/release-management-p

**Comment by kealjones-wk** [comment]
@Workiva/release-management-p

**Comment by maxwellpeterson-wf** [comment]
+1 security refresh

**Pull request merged**